### PR TITLE
Btree

### PR DESCRIPTION
**Pull request by krux02**
* performance improvements: binary search within nodes in the BTree
  * code refactoring:
    - use rotateLeft from algorithm where possible (replaces loops)
    - search for nodes is now binary
  * delte operation for BTree, non balancing, but empty nodes get removed.
 

I did some performance tests. The binary search does help, but the BTree is much slower than I expected. The biggest improvement that I measured is when the node size is reduced from 512 to 32.

### CODE REVIEW

File: compiler/semtempl.nim [hunk 1]
```diff
@@ -560,7 +560,7 @@ proc semTemplateDef(c: PContext, n: PNode): PNode =
     s = semIdentVis(c, skTemplate, n.sons[0], {})
 
   if s.owner != nil and sfSystemModule in s.owner.flags and
-      s.name.s in ["!=", ">=", ">", "incl", "excl", "in", "notin", "isnot"]:
+      s.name.s in ["!=", ">=", ">", "incl", "excl", "in", "notin", "isnot", "..<"]:
```
Comment (Araq) [inline compiler/semtempl.nim]: Yay, unrelated commits.

Comment (krux02) [reply compiler/semtempl.nim]: Well it is not unrelated, ``..<`` is used a lot in the code and I needed this change to debug what I was working on. I simply don't want to continue debugging further problems in this branch without this fix being active, it is likely to be important.

Comment (Araq) [reply compiler/semtempl.nim]: ok.

**Comment by b3liever** [comment]
Is it going to be moved to the stdlib? And also what was the benchmark results?

**Comment by krux02** [comment]
@b3liever I did some benchmarks. And it seems that BTree's performance is poor compared to what we already have with hash maps. They are meant to be used for the compiler internally only. That doesn't mean that they won't be in the standard library (yet).
